ddupdate.8
.TH DDUPDATE "8" "Last change: Apr 2022" "ddupdate" "System Administration Utilities"
.SH NAME
.P
\fBddupdate\fR - Update DNS data for dynamic IP addresses
.SH SYNOPSIS
.B ddupdate
[\fIoptions\fR]
.SH DESCRIPTION
A tool to update dynamic IP addresses typically obtained using DHCP with
dynamic DNS service providers such as changeip.com, duckdns.org or no-ip.com.
It makes it possible to access a machine with a fixed
name like myhost.duckdns.org even if the ip address changes.
.P
The tool has a plugin structure with plugins for obtaining the actual
address (typically hardware-dependent) and to update it (service
dependent).
.P
The normal usecase is to specify all commandline options in the config
file.
However, all options in this file could be overridden by actual command line
options e. g., while testing.
.P
Using service providers and possibly also firewalls requires use of
username/password credentials.
For these, either the netrc(5) or the system keyring is used.
.P
\fBddupdate\fR is distributed with systemd support to run at regular intervals,
and with NetworkManager templates to run when interfaces goes up or down. It
fully supports ipv6 addresses and also using proxies (see ENVIRONMENT).
.PP

.SH OPTIONS

Options for normal operation, typically defined in config file:

.TP 4
\fB-H, --hostname\fR <\fIhostname\fR>
Hostname to update,  typically fully qualified. Defaults to the not really
usable host.nowhere.net

.TP 4
\fB-s, --service-plugin\fR <\fIplugin\fR>
Plugin used to update the dns data for the address obtained
from the address-plugin. Defaults to dry-run, which just prints the
address. Use \fI\-\-list-services\fR to list available
plugins.

.TP 4
\fB-a, --address-plugin\fR <\fIplugin\fR>
Plugin used to obtain the actual ip address. Defaults to default-if,
which localizes the default interface using /usr/sbin/ip and uses it's
primary address. Use \fI\-\-list-addressers\fR to list available
plugins.

.TP 4
\fB-C, --auth-plugin <\fIplugin\fR>
Plugin providing authentication credentials, either \fInetrc\fR or
\fIkeyring\fR

.TP 4
\fB-v, --ip-version\fR <\fIv4\fR|\fIv6\fR|\fIall\fR>
The kind of ip addresses to register. The addresses obtained by the
address-plugin could be either v6, v4 or both. However, the actual addresses
sent to the service plugin is filtered using this option so for example
an unused ipv6 address not becomes an official address to the host.
Defaults to \fIv4\fr.

.TP 4
\fB-L, --loglevel\fR [\fIlevel\fR]
Determine the amount of logging information. \fIlevel\fR is a symbolic
syslog level: \fIerror\fR,\fIwarning, \fIinfo\fR,
or \fIdebug\fR.
It defaults to \fIwarning\fR.

.TP 4
\fB-o, --service-option\fR <\fIplugin option\fR>
Option interpreted by service plugin, documented in \fI--help <plugin>\fR.
May be given multiple times as required.
Any option on the command line will clear the list of options as of the
config file.
See PLUGIN OPTIONS.

.TP 4
\fB-O, --address-option\fR <\fIplugin option\fR>
Option interpreted by address-plugin.
See \fI\-\-service-option\fR and PLUGIN OPTIONS.

.P
Other options:

.TP 4
\fB-c, --config-file\fR <\fIpath\fR>
File containing default values for all command line options.
The path must be absolute.
An example file is distributed with the sources.
See [FILES] below.

.TP 4
\fB-f, --force\fR
Force \fBddupdate\fR to run even if the cached value is still valid.

.TP 4
\fB-e, --execute-section <section>\fR
Only run the given section in configuration file.
Use \fI\-\-list-sections\fR to list available sections.

.TP 4
\fB-h, --help [plugin]  \fR
Print help. If given a plugin argument, prints help for this plugin.

.TP 4
\fB-S, --list-services\fR
List service provider plugins.

.TP 4
\fB-A, --list-addressers\fR
List plugins providing one or more ip addresses

.TP 4
\fB-P, --list-auth-plugins\fR
List plugins for storing credentials like \fInetrc\fR and \fIkeyring\fR.

.TP 4
\fB-E, --list-sections\fR
List available sections in configuration file.

.TP 4
\fB-V, --version\fR
Print \fBddupdate\fR version.

.SH PLUGIN OPTIONS
The plugin options are generally just handed to the plugins without
any further interpretation.
An option is either a single keyword or a \fIkey=value\fR
string.
No whitespace is allowed in \fIkey\fR or \fIvalue\fR.

.SH PLUGIN LOADING
\fBddupdate\fR looks for a directory named \fIplugins\fR and tries to load
plugins from all files in this directory. The search for \fIplugins\fR
is done, in descending priority:
.IP \(bu 4
The directory \fIplugins\fR in the same directory as the main.py module.
This is the development case, and the highest priority.
.IP \(bu 4
User plugins are searched in \fI~/.local/share/ddupdate/plugins\fR.
Setting the XDG_DATA_HOME environment relocates this to
\fI$XDG_DATA_HOME/ddupdate/plugins\fR
.IP \(bu 4
The directories listed in the XDG_DATA_DIRS environment variable,
by default \fI/usr/local/share:/usr/share\fR, are searched for
\fIddupdate/plugins\fR.

.SH EXAMPLES
.P
Please note that the command line options are normally stored
in \fI~/.config/ddupdate.conf\fR, allowing an invocation without command line
options.
.P
Update on dyndns.com using the external address as seen from the
internet, displaying the address used:

.nf
	ddupdate -a default-web-ip -s dtdns.com -H myhost.dyndns.org -l info
.fi
.P
Make a debug run without actually updating, displaying the address
on the local, default interface:

.nf
	ddupdate -a default-if -s dry-run --loglevel info -H host.dyndns.org
.fi

.SH ENVIRONMENT
\fBddupdate\fR respects the data paths defined by freedesktop.org.

.TP 4
.B XDG_CACHE_HOME
Locates the cached addresses files. See FILES.

.TP 4
.B XDG_DATA_HOME
Locates user plugins. See PLUGIN LOADING.

.TP 4
.B XDG_DATA_DIRS
Involved in system plugins, see PLUGIN LOADING.

.TP 4
.B XDG_CONFIG_HOME
User configuration file parent directory location, defaults to
\fI~/.config\fR.
.P
\fBddupdate\fR also accepts the standard proxy environment:

.TP 4
.B http_proxy, https_proxy
URL to used proxies for http and https connections. The systemd service
files distributed has provisions to define these as required.


.SH FILES
.TP 4
.B ~/.netrc
When configured with the \fInetrc\fR authentication backend, this file is
used to store username and password for logging in to service providers.
See netrc(5) for the format used.
The file must have restricted permissions like 600 to be accepted.
.TP 4
.B /etc/netrc
Fallback location for credentials when \fI~/.netrc\fR is not found.
The use of this file is deprecated.
.TP 4
.B ~/.config/ddupdate.conf
Default config file location. If defined, the XDG_CONFIG_HOME variable
relocates this to \fI$XDG_CONFIG_HOME/ddupdate.conf\fR.
.TP 4
.B  /etc/ddupdate.conf
Fallback configuration file location.
.TP 4
.B /usr/share/ddupdate/plugins
Default directory for upstream plugins, see PLUGIN LOADING.
.TP 4
.B /usr/local/share/ddupdate/plugins
Default directory for site plugins, see PLUGIN LOADING.
.TP 4
.B ~/.local/share/ddupdate/plugins
Default directory for user plugins, see PLUGIN LOADING.
.TP 4
.B  ~/.cache/ddupdate/*
Cached address from last update with an actual change, one for each
update service.
Setting the XDG_CACHE_HOME environment variable relocates these files to
$XDG_CACHE_HOME/ddupdate/*.

.SH "SEE ALSO"
.TP 4
.B ddupdate.conf(5)
Configuration file
.TP 4
.B ddupdate-config(8)
Configuration tool
.TP 4
.B netrc(5)
Authentication tokens file, originally used by ftp(1), used by the
netrc authentication backend.
.TP 4
.B https://pypi.org/project/keyring/
Interface for the keyring authentication backend
.TP 4
.B  https://github.com/leamas/ddupdate
Project homesite and README



NEWS
0.7.1
* Drop the dnsdynamic plugin, service is discontinued.
* Drop the system-ns plugin, service is discontinued.
* Add new plugin for desec.io (#36).
* Add new plugin for DNS-O-Matic (#57)
* Handle expired certificate on duiadns server.
* myonlineportal, dnsexit, dynu: Handle API updates.
* Multiple bugfixes in new credentials backends.
* Base64-encode passwords in auth_netrc to handle 'strange' characters (#49).
* Add support for selecting auth plugin to ddupdate-config.
* ddplugin: Add a new header kwarg argument to get_response(), used in
  desec.io to allow token based header authentication.
* Add pyproject.toml, first step in distutils being dropped in 3.12
* Various documentation polish.
* Rename all address plugins using an addr_ prefix.
* Fix pylint and pycodestyle diagnostics.


0.7.0 
* Support multiple configuration sections in config file (#25)
* Experimental support for keyring password storage (#23)
* Add FINAL_PREFIX environment variable support to ease packaging
* Support multiple credentials for Google domains (#52)
* New address plugin for OnHub/Google WiFi/Nest WiFi  (#58)
* New dnshome.de address- and service plugins  (#47)
* Plugins: freedns: Update to simplified https API.
Thanks to m-jung, Alexey Grevtsev, Marcin Mielniczuk, Gabriel de Perthuis

0.6.6
* Fix wrong path in systemd service file (#54).
* Fix bad return value causing systemd problems (#53).
* Graceful exist on missing python-requests (#48).
Thanks to Teemo Ikonen

0.6.5
* Fix ip route parsing (#35)
* Fix python 3.9 syntax warnings (#38, #37)
* New plugin domains.google.com -- googledomains.py
* Error logging bugfixes (#4411454).
Thanks to aerusso, LoganK and Hroncok

0.6.4
* Fix address-options and service-options parsing regression (#24).
* Add new nsupdate(1) plugin (#29).
* Install systemd files in correct user locations (#22).
* Enhance ipv6 address parsing in address plugins (#28).
* Patch ddupdate.service script path on installation.
* Thanks to aerusso for nsupdate plugin, bugfixes and enhancements.

0.6.3
* Fix #21, broken ipv6 configuration file  parsing.
* Documentation bugfixes.

0.6.2
* New cloudflare plugin, thanks to Mika Mannermaa <mmannerm@gmail.com> 
* New  dy.fi, restricted ti what looks like finnish addresses.
  Thanks to Teemu Ikonen <tpikonen@gmail.com>
* Fix bug causing frequent updates due to cache misses (#19).
  Kudos: Teemu Ikonen
* Code and documentation cleanup.


0.6.1
* Fix crash when logging updates disabled by caching (#11).
* afraid.org: Use now available https url (#12).
* afraid.org: Support user-supplied addresses (#13).
* Misc cleanup.

0.6.0
* New setup by ddupdate-config based on a user service. Nothing installed
  as root, service runs as invoking user. Existing setups based on a
  separate user running service continues to work, but must be maintained
  manually.
* New service plugin dnspark.com
* New Hurricane Electric service plugin.
* New pkg-config build dependency.

9.5.3
* Remove the straight.plugins dependency.
* config: Fix bug creating new config files like ~/.config/dduodate.conf
* config: Add new address plugin option "Use address as of service".
* build: Don't install documentation files (bad idea from beginning).
* Miscellanaous cleanup and doc fixes.

0.5.2
* Multiple bugfixes in config script
* Add missing commands and short options to completions
* Move completions to proper /usr/share/bash-comletions/completions
* Drop sudo usage, use plain su instead.

0.5.1
* Add bash completion support
* Fix a bug in the user config file ~/.config/ddupdate.conf path.
* Fix Respect XDG_CONFIG_HOME (#8).
* Fix handle .netrc permissions correct (#7).
* Clean up the Makefile build-install-clean targets.
* Documentation re-organized and updated.

0.5.0
* Add ddupate-config configuration script.
* Add missing ddupdate.conf.5 manpage
* Static code checkers cleanup
* Documentation updates
* Fix bogus error message in --list-services/addressers.

0.4.1
* Documentation fixes

0.4.0:
* Declared as beta state.
* Revise API

0.3.0
* Split old --list command to --list-services and list-addressers
* Split old --options to --address-options and --service-options.
* Incompatible CLI changes
* Documentation fixes.

0.2.1:
* Fix start exception on python 3.4/jessie (#6).
* Fix unhelpful error messages for bad .netrc (#5).
* Documentation fixes.

0.2.0:
* Revise and finish ipv6 support (#3)
* New option --version, ipv4/ipv6/all addresses switch.
* New ip plugin default-web-ip6, external ipv6 address.
* Update and document proxy usage (#4)

0.1.0:
* Multitude of minor bugfixes.
* API: Finalizing current revision. Renaming silly
  named plugins_base -> ddplugin.
* Adding pydocstyle checks and updating source
* Review and bugfixes for package generation
" Added COPR and Ubuntu PPA downstream repos.

0.0.6:
* Fixed several bugs in plugin load paths.
* Fixed a bug in config file path computation.
* Removed the generated file ddupdate.8.html from distribution
* Added timeout handling in plugin http handling.
* New plugin hardcoded-ip
* Minor bugfixes in both plugins and central code.
* Added fast-track configuration to README.md, clean up.
* Cleaned up and refactored code in main module
* Fixed a bug when symlinking the ddupdate script.

0.0.5:
* Added a NEWS file
* Revised and cleaned up plugin interface.
* Added ipv6 support.
* New plugins myonlineportal.com and dynv6.com
* Fedora and debian packaging verified.


ddupdate.py

''' Update DNS data for dynamically ip addresses  '''

import argparse
import configparser
import logging
import math
import os
import os.path
import stat
import sys
import time

from straight.plugin import load

from plugins.plugins_base import IpPlugin, IpLookupError
from plugins.plugins_base import UpdatePlugin, UpdateError


if 'XDG_CACHE_HOME' in os.environ:
    CACHE_DIR = os.environ['XDG_CACHE_HOME']
else:
    CACHE_DIR = os.path.expanduser('~/.cache')

DEFAULTS = {
    'hostname': 'host.nowhere.net',
    'ip-plugin': 'default-if',
    'service-plugin': 'dry-run',
    'loglevel': 'info',
    'options': None,
    'ip-cache': os.path.join(CACHE_DIR, 'ddupdate'),
    'force': False
}


def ip_cache_setup(opts):
    ''' Ensure that our cache directory exists, return cache file path '''
    if not os.path.exists(os.path.dirname(opts.ip_cache)):
        os.makedirs(os.path.dirname(opts.ip_cache))
    return os.path.join(opts.ip_cache, opts.service_plugin + '.ip')


def ip_cache_clear(opts, log):
    ''' Remove the cache file for actual service plugin in opts. '''
    path = ip_cache_setup(opts)
    if not os.path.exists(path):
        return
    log.debug("Removing cache file: " + path)
    os.unlink(path)


def ip_cache_data(opts, default=("0.0.0.0", 100000)):
    ''' Return  a (address, cache age in minute) tuples. If not existing,
    the default value is returned.
    '''
    path = ip_cache_setup(opts)
    if not os.path.exists(path):
        return default
    mtime = os.stat(path)[stat.ST_MTIME]
    now = time.time()
    delta = math.floor((now - mtime) / 60)
    with open(path) as f:
        addr = f.read().strip()
    return addr, delta


def ip_cache_set(opts, addr):
    ''' Set the cached address to string addr. '''

    path = ip_cache_setup(opts)
    addr = addr if addr else "0.0.0.0"
    with open(path, "w") as f:
        f.write(addr.strip())


def here(path):
    ' Return path added to current dir for __file__. '
    return os.path.join(os.path.dirname(os.path.abspath(__file__)), path)


def parse_conffile(log):
    ' Parse config file path, returns verified path or None. '
    path = '/etc/ddupdate.conf'
    for i in range(len(sys.argv)):
        arg = sys.argv[i]
        if arg.startswith('-c') or arg.startswith('--conf'):
            if arg.startswith('-c') and len(arg) > 2:
                path = arg[2:]
            elif '=' in arg:
                path = arg.split('=')[1]
            elif i < len(sys.argv) - 1:
                path = sys.argv[i + 1]
            else:
                # Trust that the regular parsing handles the error.
                return None
    if not os.access(path, os.R_OK):
        log.warning("Cannot open config file '%s' for read", path)
        return None
    return path


def parse_config(path, log):
    ' Parse config file, return fully populated dict of key-values '
    results = {}
    config = configparser.ConfigParser()
    config.read(path)
    if 'update' not in config:
        log.warning(
            'No [update] section found in %s, file ignored', path)
        items = {}
    else:
        items = config['update']
    for key in DEFAULTS:
        if key not in items:
            results[key] = DEFAULTS[key]
        else:
            results[key] = items[key]
    return results


def get_parser(conf):
    ''' Construct the argparser. '''
    parser = argparse.ArgumentParser(
        prog='ddupdate',
        add_help=False,
        description="Tool to update DNS data for dynamic ip addresses")
    normals = parser.add_argument_group()
    normals.title = "Normal operation options"
    normals.add_argument(
        "-H", "--hostname", metavar="host",
        help='Hostname to update [host.nowhere.net]',
        default=conf['hostname'])
    normals.add_argument(
        "-s", "--service-plugin", metavar="plugin",
        help='Service plugin used to actually update a dns hostname address',
        default=conf['service-plugin'])
    normals.add_argument(
        "-i", "--ip-plugin", metavar="plugin",
        help='Ip plugin used to obtain the ip address to use',
        default=conf['ip-plugin'])
    normals.add_argument(
        "-c", "--config-file", metavar="path",
        help='Config file with default values for all options'
        + ' [/etc/ddupdate.conf]',
        dest='config_file', default='/etc/ddupdate.conf')
    normals.add_argument(
        "-L", "--loglevel", metavar='level',
        choices=['error', 'warning', 'info', 'debug'],
        help='Amount of printed diagnostics [warning]',
        default=conf['loglevel'])
    normals.add_argument(
        "-o", "--option", metavar="plugin option",
        help='Plugin option (enter multiple times if required)',
        dest='options', action='append')
    others = parser.add_argument_group()
    others.title = "Other options"
    others.add_argument(
        "-l", "--list-plugins", metavar="kind",
        choices=['services', 'ip-plugins', 'all'],
        help='List plugins of given kind: '
        + 'ip-plugins, services or all  [all]',
        const='all', nargs='?')
    others.add_argument(
        "-f", "--force",
        help='Force run even if the cache is fresh',
        default=False, action='store_true')
    others.add_argument(
        "-h", "--help", metavar="plugin",
        help='Print overall help or help for given plugin',
        nargs='?', const='-')
    others.add_argument(
        "-v", "--version",
        help='Print ddupdate version and exit',
        action='version')
    return parser


def parse_options(conf):
    ''' Parse command line using conf as defaults, return namespace. '''
    level_by_name = {
        'error': logging.ERROR,
        'warn': logging.WARNING,
        'warning': logging.WARNING,
        'info': logging.INFO,
        'debug': logging.DEBUG,
    }
    parser = get_parser(conf)
    parser.version = "0.0.2"
    opts = parser.parse_args()
    if opts.help == '-':
        parser.print_help()
        sys.exit(0)
    if not opts.options:
        opts.options = conf['options']
    opts.loglevel = level_by_name[opts.loglevel]
    opts.ip_cache = conf['ip-cache']
    return opts


def log_setup():
    ' Setup the module log. '
    log = logging.getLogger('ddupdate')
    log.setLevel(logging.DEBUG)
    handler = logging.StreamHandler()
    handler.setLevel(logging.INFO)
    formatter = logging.Formatter("%(levelname)s - %(message)s")
    handler.setFormatter(formatter)
    log.addHandler(handler)
    return log


def log_options(log, args):
    ' Print some info on seledted options. '
    log.info("Loglevel: " + logging.getLevelName(args.loglevel))
    log.info("Using hostname: " + args.hostname)
    log.info("Using ip address plugin: " + args.ip_plugin)
    log.info("Using service plugin: " + args.service_plugin)
    log.info("Plugin options: "
             + (' '.join(args.options) if args.options else ''))


def load_plugins(path, log):
    ''' Load ip and service plugins into dicts keyed by name. '''
    path = os.path.join(path, 'plugins')
    # This is weird, likely a bug. Absolute paths does not make it
    path = os.path.relpath(path)
    log.debug("Loading plugins from: %s", path)
    if not os.path.exists(path):
        return {}, {}
    getters = load(path, subclasses=IpPlugin)
    getters = getters.produce()
    getters_by_name = {plug.name(): plug for plug in getters}
    setters = load(path, subclasses=UpdatePlugin)
    setters = setters.produce()
    setters_by_name = {plug.name(): plug for plug in setters}
    log.debug("Loaded %d address and %d service plugins",
              len(getters), len(setters))
    return getters_by_name, setters_by_name


def list_plugins(ip_plugins, service_plugins, kind):
    ''' List all loaded plugins (noreturn). '''
    if kind == 'all' or kind.startswith('i'):
        for name, plugin in sorted(ip_plugins.items()):
            print("%-20s %s" % (name, plugin.oneliner()))
    if kind == 'all' or kind.startswith('s'):
        for name, plugin in sorted(service_plugins.items()):
            print("%-20s %s" % (name, plugin.oneliner()))
    sys.exit(0)


def plugin_help(ip_plugins, service_plugins, plugid):
    ''' print full help for given plugin (noreturn).'''
    if plugid in ip_plugins:
        plugin = ip_plugins[plugid]
    elif plugid in service_plugins:
        plugin = service_plugins[plugid]
    else:
        print("No help available (nu such plugin?): " + plugid)
        sys.exit(2)
    print("Name: " + plugin.name())
    print("Source: " + plugin.sourcefile() + "\n")
    print(plugin.info())

    sys.exit(0)


def build_load_path(log):
    ''' Return list of paths to load plugins from. '''
    paths = []
    path = os.path.expanduser('~/.local/share')
    if 'XDG_DATA_HOME' in os.environ:
        path = os.environ['XDG_DATA_HOME']
    paths.append(path)
    syspaths = "/usr/local/share:/usr/share"
    if 'XDG_DATA_DIRS' in os.environ:
        syspaths = os.environ['XDG_DATA_DIRS']
    paths.extend(syspaths.split(':'))
    paths = [os.path.join(p, 'ddupdate') for p in paths]
    paths.insert(0, os.getcwd())
    log.debug('paths :%s', ':'.join(paths))
    return paths


def main():
    ''' Indeed: main function. '''
    ip_plugins = {}
    service_plugins = {}
    log = log_setup()
    conffile_path = parse_conffile(log)
    conf = parse_config(conffile_path, log) if conffile_path else DEFAULTS
    opts = parse_options(conf)
    log.handlers[0].setLevel(opts.loglevel)
    log_options(log, opts)
    load_paths = build_load_path(log)
    for path in reversed(load_paths):
        getters, setters = load_plugins(path, log)
        ip_plugins.update(getters)
        service_plugins.update(setters)
    if opts.help and opts.help != '-':
        plugin_help(ip_plugins, service_plugins, opts.help)
    if opts.list_plugins:
        list_plugins(ip_plugins, service_plugins, opts.list_plugins)
    if opts.ip_plugin not in ip_plugins:
        log.error("No such ip plugin: %s", opts.ip_plugin)
        sys.exit(2)
    try:
        ip = ip_plugins[opts.ip_plugin].run(opts, log)
    except IpLookupError as err:
        log.error("Cannot obtain ip address: %s", err)
        sys.exit(3)
    if ip == "0.0.0.0":
        log.info("Using ip address provided by update service")
        ip = None
    else:
        log.info("Using ip address: " + ip)
    if opts.force:
        ip_cache_clear(opts, log)
    addr, age = ip_cache_data(opts)
    if opts.service_plugin not in service_plugins:
        log.error("No such service plugin: %s", opts.service_plugin)
        sys.exit(2)
    service_plugin = service_plugins[opts.service_plugin]
    if age < service_plugin.ip_cache_ttl() and (addr == ip or not ip):
        log.info("Update inhibited, cache is fresh (%d min)", age)
        sys.exit(0)
    try:
        service_plugin.run(opts, log, ip)
    except UpdateError as err:
        log.error("Cannot update DNS data: %s", err)
    else:
        ip_cache_set(opts, ip)
        log.info("Update OK")


if __name__ == '__main__':
    main()


# vim: set expandtab ts=4 sw=4:
